src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/RuntimeAsyncRewriter.cs

```diff
@@ -112,4 +118,81 @@ public BoundExpression VisitExpression(BoundExpression node)
         // System.Runtime.CompilerServices.RuntimeHelpers.Await(awaitedExpression)
         return _factory.Call(receiver: null, awaitMethod, VisitExpression(node.Expression));
     }
+
+    private BoundExpression RewriteCustomAwaiterAwait(BoundAwaitExpression node)
+    {
+        // await expr
+        // becomes
+        // var _tmp = expr.GetAwaiter();
+        // if (!_tmp.IsCompleted)
+        //    UnsafeAwaitAwaiterFromRuntimeAsync(_tmp) OR AwaitAwaiterFromRuntimeAsync(_tmp);
+        // _tmp.GetResult()
+
+        // PROTOTYPE: await dynamic will need runtime checks, see AsyncMethodToStateMachine.GenerateAwaitOnCompletedDynamic
+
+        var expr = VisitExpression(node.Expression);
+
+        var awaitableInfo = node.AwaitableInfo;
+        var awaitablePlaceholder = awaitableInfo.AwaitableInstancePlaceholder;
+        if (awaitablePlaceholder is not null)
+        {
+            _placeholderMap.Add(awaitablePlaceholder, expr);
+        }
+
+        // expr.GetAwaiter()
+        var getAwaiter = VisitExpression(awaitableInfo.GetAwaiter);
+        Debug.Assert(getAwaiter is not null);
+
+        if (awaitablePlaceholder is not null)
+        {
+            _placeholderMap.Remove(awaitablePlaceholder);
+        }
+
+        // var _tmp = expr.GetAwaiter();
+        var tmp = _factory.StoreToTemp(getAwaiter, out BoundAssignmentOperator store, kind: SynthesizedLocalKind.Awaiter);
+
+        // _tmp.IsCompleted
+        Debug.Assert(awaitableInfo.IsCompleted is not null);
+        var isCompletedMethod = awaitableInfo.IsCompleted.GetMethod;
+        Debug.Assert(isCompletedMethod is not null);
+        var isCompletedCall = _factory.Call(tmp, isCompletedMethod);
+
+        // UnsafeAwaitAwaiterFromRuntimeAsync(_tmp) OR AwaitAwaiterFromRuntimeAsync(_tmp)
+        var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+        var useUnsafeAwait = _factory.Compilation.Conversions.ClassifyImplicitConversionFromType(
+            tmp.Type,
+            _factory.Compilation.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ICriticalNotifyCompletion),
+            ref discardedUseSiteInfo).IsImplicit;
```

Comment:
What makes it so the use site info can be discarded? Maybe because that the target type is supposed to come from the core library?